core/src/main/java/org/apache/calcite/rel/core/JoinInfo.java

```diff
@@ -31,23 +31,25 @@
 
 /** An analyzed join condition.
  *
- * <p>It is useful for the many algorithms that care whether a join is an
- * equi-join.
+ * <p>It is useful for the many algorithms that care whether a join has an
+ * equi-join condition.
  *
  * <p>You can create one using {@link #of}, or call
  * {@link Join#analyzeCondition()}; many kinds of join cache their
- * join info, especially those that are equi-joins and sub-class
- * {@link org.apache.calcite.rel.core.EquiJoin}.</p>
+ * join info, especially those that are equi-joins.</p>
  *
  * @see Join#analyzeCondition() */
-public abstract class JoinInfo {
+public class JoinInfo {
   public final ImmutableIntList leftKeys;
   public final ImmutableIntList rightKeys;
+  protected final RexNode nonEquiCondition;
```

Comment:
My instinct is to make everything non-null. I don't like the test `nonEquiCondition == null || nonEquiCondition.isAlwaysTrue()`, for instance.

Consider making `nonEquiCondition` a conjunctive list (immutable, final and not null); if the list is empty, that means true. We do this in several other places, and it is useful, because caller can assume that ANDs have been flattened.

Reply:
`getRemaining` is a public method and returns `RexNode`. If we make `nonEquiCondition` a conjunctive list, we have to convert to a `RexNode` every time we call `getRemainng`. How about checking the nullness of input `nonEquiCondition` in the `JoinInfo` constructor, set it to literal `TRUE` rexnode if it is `null`? At the same time, add `@Nonnull` to `getRemaining`.

Reply:
Revoke what I just said, the is no way to pass `RexBuilder` in the constructor.

Reply:
I don't think you should allow availability of `RexBuilder` to sway the design. When a `JoinInfo` is created, there is always a `RexBuilder` available, and that `RexBuilder` (and its `RelDataTypeFactory`) remain valid for the lifetime of the `JoinInfo`.

Representing conditions as conjunctive lists seems to be very convenient. For example, consider the `RelOptUtil.splitJoinCondition` method, which is the source of `JoinInfo.remaining`. The last thing it does is to convert a conjunctive list into a `RexNode`. Other callers of that method either ignore the return, or immediately convert the `RexNode` back into a conjunctive list.

Reply:
Make sense. I will change nonEquiCondition to conjunctive list. But looks like we have to keep `public RexNode getRemaining(RexBuilder rexBuilder)` unchanged, since it is a public API.